[Index Management] Support managed_by field and fix never delete data logic

sabarasaba opened the pull request:
Closes https://github.com/elastic/kibana/issues/167414
Closes https://github.com/elastic/kibana/issues/167416
Closes https://github.com/elastic/kibana/issues/167805

## Summary
This PR introduces a couple of separate states for the edit data retention modal, that can be triggered depending on the state of the datastream.

##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial -E path.repo=/tmp/snapshots` and kibana with `yarn start`
2. Navigate to console and execute the snippet for the case you wanna test
3. Navigate to index management and click on data streams tab. 
4. Verify that:
  * When the datastream is fully managed by ILM, edit option shouldnt be shown in the manage options and a little warning callout should be shown in the details flyout
  * When is fully managed by DSL, no callouts should be shown
  * When is managed by DSL, but has ILM backing indices we allow editing but show a callout saying that some indices are only managed by ILM

<details>
<summary>Create a datastream that is fully managed by ILM</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}
```
</details>

<details>
<summary>Create a datastream that is unmanaged</summary>

```
PUT _index_template/my-data-stream-template
{
    "index_patterns": [ "my-data-stream*" ],
    "data_stream": {},
    "priority": 400,
    "template" : {
        "lifecycle": { 
           "data_retention": "7d",
           "enabled": false 
        }
    }
}


POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SYKENET-000001",
  "location": "swamp",
  "temperature": 32.4,
  "humidity": 88.9
}
```
</details>

<details>
<summary>Create a datastream that managed by DSL but has ILM indices</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}



PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm",
      "index.lifecycle.prefer_ilm": false
    },
    "lifecycle": {
      "data_retention": "7d"
    }
  }
}

POST my-data-stream2/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}

PUT _data_stream/my-data-stream/_lifecycle
{
  "data_retention": "7d"
}
```
</details>

<details>
<summary>Create a datastream fully managed by DSL</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm",
      "index.lifecycle.prefer_ilm": false
    },
    "lifecycle": {
      "data_retention": "7d"
    }
  }
}

POST my-data-stream2/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}
```
</details>

<details>
<summary>Screenshots</summary>

// Fully managed by ILM
![Screenshot 2023-10-25 at 10 15 52](https://github.com/elastic/kibana/assets/1191206/f843eae4-4aa3-416b-a3c2-b68c8e2831f9)
// --> also when its fully managed by ILM the computed lifecycle is managed by ILM, so it will be strikedthrough in the table and flyout
![Screenshot 2023-10-26 at 17 25 12](https://github.com/elastic/kibana/assets/1191206/cf2a01da-5ecc-4d5e-8cea-d327ee4b3321)


// Fully managed by DSL
![Screenshot 2023-10-25 at 10 20 45](https://github.com/elastic/kibana/assets/1191206/900b6d38-7d62-4b5d-9f4f-a71b5acda3b8)
![Screenshot 2023-10-26 at 10 17 36](https://github.com/elastic/kibana/assets/1191206/03e261ed-a6ff-4a02-b724-75c075dd9528)
![Screenshot 2023-10-26 at 10 18 18](https://github.com/elastic/kibana/assets/1191206/684cbfd8-b35c-4fa6-a5db-64eecd4f73b6)



// Managed by DSL but has backing indices managed by ILM
![Screenshot 2023-10-26 at 10 16 29](https://github.com/elastic/kibana/assets/1191206/8d504dbd-2391-470b-9a19-d5610d128018)
![Screenshot 2023-10-26 at 10 16 54](https://github.com/elastic/kibana/assets/1191206/828984ee-f71e-4ee9-a423-963331b029b4)


</details>

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_detail_panel/data_stream_detail_panel.tsx:
```diff
@@ -273,15 +298,58 @@ export const DataStreamDetailPanel: React.FunctionComponent<Props> = ({
+            <EuiCallOut
+              title={i18n.translate(
+                'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.fullyManagedByILMTitle',
+                { defaultMessage: 'This data stream is fully managed by ILM' }
```
```suggestion
                { defaultMessage: 'This data stream and its associated indices are managed by ILM' }
```
Mostly trying to disambiguate the meaning of "fully" here to be as explicit as possible.

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_detail_panel/data_stream_detail_panel.tsx:
```diff
@@ -273,15 +298,58 @@ export const DataStreamDetailPanel: React.FunctionComponent<Props> = ({
+              <p>
+                <FormattedMessage
+                  id="xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.fullyManagedByILMDescription"
+                  defaultMessage="Editing the data retention for this data stream is only possible through its {link}."
```
```suggestion
                  defaultMessage="To edit data retention for this data stream, you must edit its associated {link}."
```
Minor styling edit for a more active voice.

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_table/data_stream_table.tsx:
```diff
@@ -138,7 +140,7 @@ export const DataStreamTable: React.FunctionComponent<Props> = ({
         content={i18n.translate('xpack.idxMgmt.dataStreamList.table.dataRetentionColumnTooltip', {
           defaultMessage:
-            '[Technical preview] Data will be kept at least this long before it is automatically deleted. Only applies to data streams managed by a data stream lifecycle. This value might not apply to all data if the data stream also has an index lifecycle policy.',
+            'Data will be kept at least this long before it is automatically deleted. Only applies to data streams managed by a data stream lifecycle. This value might not apply to all data if the data stream also has an index lifecycle policy.',
```
```suggestion
            'Data is kept at least this long before being automatically deleted. The data retention value only applies to the data managed directly by the data stream. If some data is subject to an index lifecycle management policy, then the data retention value set for the data stream doesn't apply to that data.',
```
Trying to make this tooltip a little more explicit, but not 100% it's a success :) Feel free to ignore or adjust.

sabarasaba replied:
love this!

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -146,13 +151,94 @@ const configurationFormSchema: FormSchema = {
+    label: i18n.translate(
+      'xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.dataRetentionEnabledField',
+      {
+        defaultMessage: 'Enable data retention',
```
```suggestion
        defaultMessage: 'Enable data retention',
```
What does this do when it's disabled? No data is kept? Or there's still a minimum?
Wondering:
- if it's useful to show a little help text under the toggle to be explicit about what happens to the data when it's disabled (it might be obvious but I don't know this part very well ^^')
- What's the "value" we show in the flyout and table under Data retention when it's disabled? Does it hint on what happens to the data (if useful)? Sorry I tried to test it locally but my Kibana shows a blank screen after the loading screen.

sabarasaba replied:
When its disabled we show in the UI just a text saying `Disabled`, and when its infinite data retention we show a value of `∞`. I'll ask around whats the default behaviour when its disabled, as I'm not 100% sure

sabarasaba replied:
I'll go ahead and merge this work in, will follow up with you once i get an answer!

florent-leborgne commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/edit_data_retention_modal/edit_data_retention_modal.tsx:
```diff
@@ -146,13 +151,94 @@ const configurationFormSchema: FormSchema = {
+      <p>
+        <FormattedMessage
+          id="xpack.idxMgmt.dataStreamsDetailsPanel.editDataRetentionModal.someManagedByILMBody"
+          defaultMessage="The following indices are managed by ILM: {managedByILM} {hasManyIndices, plural, one {and {remainingIndices} more} other { } } ({viewAllIndicesLink}). Updating data retention will not affect these indices, instead you will have to update the {ilmPolicyLink} policy."
```
```suggestion
          defaultMessage="The following indices are managed by ILM: {managedByILM} {hasManyIndices, plural, one {and {remainingIndices} more} other { } } [{viewAllIndicesLink}]. Updating data retention for this data stream won't affect these indices. Instead you will have to update the {ilmPolicyLink} policy."
```
Do we control labels of indices or can they be 100 characters long? In case it can be randomly long, maybe 2 alternatives to consider to have more control over readability for this callout:
- Don't display index names by default, and instead say: `One or several indices are managed by an ILM policy [view indices]. Updating data [...].`. I'm also thinking that this method might be better in case there would be multiple ILM policies managing different indices (not sure it's possible, but in case).
- Have a bullet list with the first 2 or 3 ones, and have a last bullet that contains the [view all indices] link. It'll make the callout longer, but easier to read.

sabarasaba replied:
Thanks for the suggestion, I neded up going for the first option as the modal is already a big large to accomodate a bulleted list.

florent-leborgne replied:
👍 That was my favorite option too :)

florent-leborgne reviewed (commented):
Hope my comments or questions make some kind of sense :) Thanks!

yuliacech commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_table/data_stream_table.tsx:
```diff
@@ -151,7 +153,14 @@ export const DataStreamTable: React.FunctionComponent<Props> = ({
-    render: (lifecycle: DataStream['lifecycle']) => getLifecycleValue(lifecycle, INFINITE_AS_ICON),
+    render: (lifecycle: DataStream['lifecycle'], dataStream) => (
+      <ConditionalWrap
+        condition={isDataStreamFullyManagedByILM(dataStream)}
```
Do you think there is any concern in terms of performance? If I understand correctly, we are checking every index of every data stream here.

sabarasaba replied:
Good call! I've wrapped the computed property around a useMemo now

yuliacech reviewed (commented):
Great work improving this UI, @sabarasaba! It's such a complex concept, especially when both ILM and DSL are involved. 
While testing, I was not able to replicate "manage by DSL but has ILM indices", there was no warning displayed in the edit retention modal after executing the commands from the PR description. Though I see the ILM policy (with strike through) and data retention in the details panel. I think for any combination that both systems are displayed in the panel, we should display some kind of a warning/help callout. Otherwise, for me it was not clear why the ILM policy is striked through. 
<img width="491" alt="Screenshot 2023-10-27 at 17 49 36" src="https://github.com/elastic/kibana/assets/6585477/c60eab28-0e9f-4950-a30a-13d1af93a6b6">

sabarasaba commented:
Thanks for having a look @yuliacech! I've added a small tooltip to the policy when it doesnt take effect and changed the strikethrough to a dimmer text color to indicate that it just doenst take effect. This just happens when the policy has an attached ilm policy but its not managed by ILM and doenst have any backing indices that are managed by ilm anymore.

![Screenshot 2023-10-31 at 09 37 52](https://github.com/elastic/kibana/assets/1191206/1599ba45-9276-4199-8e5a-c3525aae9d19)

Strange that the above snippet didnt work, with this one right here you should end up with two data streams one that has mixed ilm/dsl and one fully managed by dsl.

<details>
<summary>Console snippet mixed ilm dsl indices</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}



PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm",
      "index.lifecycle.prefer_ilm": false
    },
    "lifecycle": {
      "data_retention": "7d"
    }
  }
}

POST my-data-stream2/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}

PUT _data_stream/my-data-stream/_lifecycle
{
  "data_retention": "7d"
}
```
</details>

yuliacech reviewed (approved):
Thanks a lot for addressing my comments, @sabarasaba! Latest changes LGTM 👍 
The greyed out ILM policy name makes sense and I was able to see the "warning" callout about some indices being managed by ILM in the modal. I still think that in this case the callout or a shorter version of it could be displayed in the flyout itself. Because otherwise we display both DSL and ILM information in the flyout at the same time without any explanation. But we can discuss this further in the DSL sync.
